src/Compilers/VisualBasic/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncMethodToClassRewriter.Spilling.vb

```diff
@@ -231,23 +235,23 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                 Return newArgs.AsImmutableOrNull
             End Function
 
-            Private Function SpillValue(expr As BoundExpression, <[In], Out> ByRef builder As SpillBuilder) As BoundExpression
-                Return SpillValue(expr, isReceiver:=False, builder:=builder)
+            Private Function SpillValue(expr As BoundExpression, shouldCheckSideEffects As Boolean, sideEffectsAlreadyChecked As Boolean, <[In], Out> ByRef builder As SpillBuilder) As BoundExpression
+                Return SpillValue(expr, isReceiver:=False, shouldCheckSideEffects, sideEffectsAlreadyChecked, builder:=builder)
             End Function
 
-            Private Function SpillValue(expr As BoundExpression, isReceiver As Boolean, <[In], Out> ByRef builder As SpillBuilder) As BoundExpression
+            Private Function SpillValue(expr As BoundExpression, isReceiver As Boolean, shouldCheckSideEffects As Boolean, sideEffectsAlreadyChecked As Boolean, <[In], Out> ByRef builder As SpillBuilder) As BoundExpression
                 If Unspillable(expr) Then
                     Return expr
 
                 ElseIf isReceiver OrElse expr.IsLValue Then
-                    Return SpillLValue(expr, isReceiver, builder)
+                    Return SpillLValue(expr, isReceiver, shouldCheckSideEffects, sideEffectsAlreadyChecked, builder)
 
                 Else
                     Return SpillRValue(expr, builder)
                 End If
             End Function
 
-            Private Function SpillLValue(expr As BoundExpression, isReceiver As Boolean, <[In], Out> ByRef builder As SpillBuilder) As BoundExpression
+            Private Function SpillLValue(expr As BoundExpression, isReceiver As Boolean, shouldCheckSideEffects As Boolean, sideEffectsAlreadyChecked As Boolean, <[In], Out> ByRef builder As SpillBuilder) As BoundExpression
```

Comment:
>shouldCheckSideEffects As Boolean, sideEffectsAlreadyChecked As Boolean [](start = 89, length = 71)

It feels like the two parameters are confusing and unnecessary complicate things. Could we get away with just one parameter, perhaps called "isAssignmentTarget" as in C#?

Reply:
Nope. This is called from multiple code paths, not just assignments to fields. We require 3 state logic here: should we cause side effects, and if so have we already done something that will trigger these side effects recursively.

Reply:
I could change this to an enum if that would be clearer?

Reply:
> Nope. This is called from multiple code paths, not just assignments to fields. We require 3 state logic here: should we cause side effects, and if so have we already done something that will trigger these side effects recursively. 

I still think there is a way to simplify this. It looks like we only want to customize side-effects evaluation for an assignment case. 
- Let's have just one parameter, ```evaluateSideEffects```
- VisitBinaryOperator passes True for it. 
- VisitLoweredConditionalAccess passes True for it. 
- SpillArgumentListInner passes True for it. 
- SpillValue overloads above pass the value through
- In this function ```Case BoundKind.Sequence``` and ```Case BoundKind.SpillSequence```  just pass it through
- ```Case BoundKind.ArrayAccess``` evaluates side-effects only if the parameter is True
- ```Case BoundKind.FieldAccess```  evaluates side-effects only if the parameter is True and the receiver is not known to be a value type. 
- It also passes True if ```evaluateSideEffects``` was true and it is not going to evaluate side-effects, False otherwise.
- ProcessRewrittenAssignmentOperator passes false and evaluates side-effects itself as follows:
Starting with the node ProcessRewrittenAssignmentOperator gets back, if the node is a field access and its receiver is not known to be a value type, it evaluates the field access and stops. Otherwise, if the node is a field access, the process is repeated for the receiver.  

Would this work?

---
In reply to: [406469496](https://github.com/dotnet/roslyn/pull/42807#discussion_r406469496) [](ancestors = 406469496)

Reply:
Perhaps we need to skip one field access in ProcessRewrittenAssignmentOperator for the purpose of side-effect evaluation. I think you should get the idea.

---
In reply to: [406508448](https://github.com/dotnet/roslyn/pull/42807#discussion_r406508448) [](ancestors = 406508448,406469496)

Reply:
I've gone for a slightly different refactoring that I think should be easier to follow without duplicating the logic for checking side effects.

Reply:
> I've gone for a slightly different refactoring that I think should be easier to follow without duplicating the logic for checking side effects.

I think this is much better.

---
In reply to: [406730966](https://github.com/dotnet/roslyn/pull/42807#discussion_r406730966) [](ancestors = 406730966)